File: common/src/main/java/dev/obscuria/fragmentum/service/FactoryService.java
package dev.obscuria.fragmentum.service;

import com.mojang.serialization.MapCodec;
import net.minecraft.core.particles.ParticleOptions;
import net.minecraft.core.particles.ParticleType;
import net.minecraft.core.particles.SimpleParticleType;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.world.entity.ai.sensing.Sensor;
import net.minecraft.world.entity.ai.sensing.SensorType;
import org.jetbrains.annotations.ApiStatus;

import java.util.function.Supplier;

@ApiStatus.Internal
public interface FactoryService {

    <T extends ParticleOptions> ParticleType<T> newParticleType(boolean alwaysSpawn, MapCodec<T> codec, StreamCodec<? super RegistryFriendlyByteBuf, T> streamCodec);

    SimpleParticleType newParticleType(boolean alwaysSpawn);

    <T extends Sensor<?>> SensorType<T> newSensorType(Supplier<T> factory);
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/service/FactoryService.kt
package dev.obscuria.fragmentum.service

import com.mojang.serialization.Codec
import net.minecraft.core.BlockPos
import net.minecraft.core.particles.ParticleOptions
import net.minecraft.core.particles.ParticleType
import net.minecraft.core.particles.SimpleParticleType
import net.minecraft.world.entity.ai.sensing.Sensor
import net.minecraft.world.entity.ai.sensing.SensorType
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import java.util.function.BiFunction
import java.util.function.Supplier

interface FactoryService {

    fun <T : BlockEntity> newBlockEntityType(
        factory: BiFunction<BlockPos, BlockState, T>,
        vararg blocks: Block
    ): BlockEntityType.Builder<T>

    fun <T : ParticleOptions> newParticleType(
        alwaysSpawn: Boolean,
        codec: Codec<T>,
        deserializer: ParticleOptions.Deserializer<T>
    ): ParticleType<T>

    fun newParticleType(alwaysSpawn: Boolean): SimpleParticleType

    fun <T : Sensor<*>?> newSensorType(factory: Supplier<T>): SensorType<T>
}